py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -270,21 +301,133 @@ def _create_noise_channel(self, noise_config: BitBoxNoiseConfig) -> NoiseConnect
             noise_config.add_device_static_pubkey(remote_static_key)
         return noise
 
-    def close(self) -> None:
-        self._transport.close()
+    def noise_connect(self, noise_config: BitBoxNoiseConfig) -> None:
+        self._noise = self._create_noise_channel(noise_config)
+
+    @abstractmethod
+    def unlock_query(self) -> bytes:
+        """
+        Executes an unlock query.
+        Returns the bytes containing the response status.
+        """
+        ...
+
+
+class BitBoxProtocolV1(BitBoxProtocol):
+    """ BitBox Protocol from firmware V1.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        raise NotImplementedError("unlock_query is not supported in BitBox protocol V1")
+
+    def _encode_noise_request(self, encrypted_msg: bytes) -> bytes:
+        return encrypted_msg
+
+    def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
+        """
+        Until V7 of the protocol, we don't encapsulate OP_NOISE_MSG responses.
+        Let's assume that if a response is only one byte long, that means it
+        contains an error.
+        """
+        if len(encrypted_msg) == 1:
+            return RESPONSE_FAILURE, b""
+        return RESPONSE_SUCCESS, encrypted_msg
+
+    def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
+        """
+        V1-6 of the BB noise protocol doesn't encapsulate handshake requests, and don't
+        send back a status code in the response.
+        """
+        noise_result = self._raw_query(req)
+        return RESPONSE_SUCCESS, noise_result
+
+class BitBoxProtocolV2(BitBoxProtocolV1):
+    """ BitBox Protocol from firmware V2.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        unlock_result, unlock_data = self.query(OP_UNLOCK, b"")
+        if len(unlock_data) != 0:
+            raise ValueError(f"OP_UNLOCK (V3) replied with wrong length.")
+        return unlock_result
```

Comment:
in v2 there is no unlock_result either. I think the signature of `unlock_query` should probably change, as not all versions have a result (v2.0.0 has no result). It could be more readable and correct if each versions `unlock_query` checked for the expected result and raised an exception otherwise, instead of passing it upwards.

Reply:
Fixed in #362